File: 236_lowest_common_ancestor_of_a_binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private void getNodePath(TreeNode root, TreeNode node, LinkedList<TreeNode> path) {
        if (root == null || path.peekLast() == node) return;
        path.addLast(root);

        getNodePath(root.left, node, path);
        getNodePath(root.right, node, path);
        if (path.peekLast() != node) {
            path.removeLast();
        }
    }

    public TreeNode getLowestCommonNode(LinkedList<TreeNode> l1, LinkedList<TreeNode> l2) {
        TreeNode node = null;

        Iterator it1 = l1.iterator();
        Iterator it2 = l2.iterator();
        TreeNode node1 = null, node2 = null;

        while (it1.hasNext() || it2.hasNext()) {
            if (it1.hasNext()) {
                node1 = (TreeNode)it1.next();
            }
            if (it2.hasNext()) {
                node2 = (TreeNode)it2.next();
            }
            if (node1 != null && node2 != null && node1.val == node2.val) {
                node = node1;
            }
        }
        return node;
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || p == null || q == null) return null;

        LinkedList<TreeNode> listP = new LinkedList<>();
        getNodePath(root, p, listP);

        LinkedList<TreeNode> listQ = new LinkedList<>();
        getNodePath(root, q, listQ);

        TreeNode res = getLowestCommonNode(listP, listQ);
        return res;
    }

    private void printList(LinkedList<TreeNode> list) {
        Iterator iterator = list.iterator();
        while (iterator.hasNext()) {
            System.out.print(((TreeNode)iterator.next()).val + " ");
        }
        System.out.println();
    }
}

File: python/236_lowest_common_ancestor_of_a_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def getPath(root, target_node, res):
            if not root or (res and res[-1] == target_node): return

            res.append(root)
            getPath(root.left, target_node, res)
            getPath(root.right, target_node, res)
            if res[-1] != target_node:
                res.pop()

        def getLowestCommonNode(l1, l2):
            iter1, iter2 = iter(l1), iter(l2)
            res = None

            while True:
                node_1 = next(iter1, None)
                node_2 = next(iter2, None)
                if not node_1 and not node_2:
                    break
                if node_1 and node_2 and node_1.val == node_2.val:
                    res = node_1
            return res

        if not root or not p or not q: return None

        list_p = []
        getPath(root, p, list_p)

        list_q = []
        getPath(root, q, list_q)

        return getLowestCommonNode(list_p, list_q)
